File: _includes/code/automated-testing.py
# Blog: Automated testing for Weaviate applications - Python code
#SATRT Connect  # START ConnectAndCleanup
import weaviate
import os

# Instantiate the embedded Weaviate client and pass the OpenAI API key
client = weaviate.Client(
    embedded_options=weaviate.embedded.EmbeddedOptions(),
    additional_headers={
        'X-OpenAI-Api-Key': os.environ['OPENAI_API_KEY']  # Replace with your OpenAI API key
    }
)
# Client is now ready to accept requests

# Clean slate for local testing (GitHub Actions VMs start fresh) because Weaviate data is persisted in embedded mode.
client.schema.delete_all()
# Client connected and schema cleared


# Create the class
class_name = 'JeopardyQuestion'
class_definition = {
    'class': class_name,
    'vectorizer': 'text2vec-openai',
}

client.schema.create_class(class_definition)

# Test
retrieved_definition = client.schema.get(class_name)
assert retrieved_definition['moduleConfig']['text2vec-openai']['model'] == 'ada'
# Class created successfully


# Import objects from the JSON file
import json

with open('jeopardy_100.json') as f:
    data = json.load(f)

# Context manager for batch import
with client.batch as batch:
    # Build data objects & add to batch
    for i, obj in enumerate(data):
        batch.add_data_object(
            data_object={
                'question': obj['Question'],
                'answer': obj['Answer'],
            },
            class_name=class_name,
        )
    if i % 20 == 0:
        print(f'Imported {i} objects...')

# Test all objects were imported
response = client.query.aggregate(class_name).with_meta_count().do()
actual_count = response['data']['Aggregate'][class_name][0]['meta']['count']
assert actual_count == 100, f'Expected 100 imported objects but got {actual_count}'
# Import completed successfully


# Run a test query
result = (
    client.query
    .get('JeopardyQuestion', ['question', 'answer'])
    .with_near_text({'concepts': 'chemistry'})
    .with_limit(1)
    .do()
)

# Test
assert 'sodium' in result['data']['Get'][class_name][0]['answer']
# Query test completed

File: _includes/code/automated-testing.ts
// Blog: Automated testing for Weaviate applications - TypeScript code
// run with: node --loader=ts-node/esm automated-testing.ts

// START Connect  // START ConnectAndCleanup
import weaviate, { EmbeddedOptions } from 'weaviate-ts-embedded';
import assert from 'assert';

// Instantiate the embedded Weaviate client and pass the OpenAI API key
const client = weaviate.client(new EmbeddedOptions(
  {
    env: {
      OPENAI_APIKEY: process.env['OPENAI_API_KEY'],
    },
  }
));

await client.embedded.start();
// Client is now ready to accept requests

const className = 'JeopardyQuestion';

// Clean slate for local testing (GitHub Actions VMs start fresh) because Weaviate data is persisted in embedded mode.
try {
  await client.schema.classDeleter().withClassName(className).do();
} catch {
  // ignore error if class doesn't exist
}
// Client connected and schema cleared


// Create the class
const classDefinition = {
  class: className,
  vectorizer: 'text2vec-openai',
};

// The return value is the class, as returned by client.schema.classGetter().withClassName(className).do();
const retrievedDefinition = await client.schema.classCreator().withClass(classDefinition).do();
// Test
assert.equal(retrievedDefinition.moduleConfig['text2vec-openai']['model'], 'ada');
// Class created successfully

// Import objects from the JSON file
import data from './jeopardy_100.json' assert { type: 'json' };

let batcher = client.batch.objectsBatcher();
let counter = 0;
const batchSize = 20;

async function addObject(obj: object): Promise<void> {
  // Add the object to the batch queue
  batcher = batcher.withObject({
    class: className,
    properties: {
      question: obj['Question'],
      answer: obj['Answer'],
    },
  });

  // When the batch counter reaches batchSize, push the objects to Weaviate
  if (++counter % batchSize === 0) {
    // Flush the batch queue and restart it
    const response = await batcher.do();
    batcher = client.batch.objectsBatcher();

    // Handle errors
    for (const r of response)
      if (r.result.errors)
        throw r.result.errors;

    console.log(`Imported ${counter} articles...`);
  }
}

for (const object of data) {
  await addObject(object);
}

// Flush any remaining objects
if (batcher.payload().objects.length > 0)
  await batcher.do();

console.log(`Finished importing ${counter} objects.`);

// Test all objects were imported
const result = await client.graphql.aggregate().withClassName(className).withFields('meta { count }').do();
assert.deepEqual(result.data['Aggregate'][className], [{ meta: { count: 100 } }]);
// Import completed successfully

// Run a test query
const results = await client.graphql
  .get()
  .withClassName(className)
  .withNearText({ concepts: ['chemistry'] })
  .withLimit(1)
  .withFields('question answer')
  .do();

// Test
assert(results.data.Get[className][0]['answer'].includes('sodium'));
// Query test completed

await client.embedded.stop();
// END all
